File: backend/app/agents/intelligence.py
"""
Intelligence layer — pure math functions for scoring, detection, and segmentation.
Server-side port of frontend/lib/intelligence.ts.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone


@dataclass
class ScoredProduct:
    id: str
    title: str
    handle: str
    status: str
    price: float
    inventory: int
    score: int
    tier: str  # Core, Strong, Slow, Exit
    velocity: float
    days_left: int
    trend: str  # growing, stable, declining
    trend_ratio: float
    revenue_total: float
    image: str | None = None


@dataclass
class HealthIssue:
    product_id: str
    product_title: str
    issue: str
    severity: str  # critical, warning, info


@dataclass
class DiscountSuggestion:
    product: ScoredProduct
    discount_pct: int


def _power_scale(value: float, max_value: float) -> int:
    if value <= 0 or max_value <= 0:
        return 0
    return min(100, round(math.pow(value / max_value, 0.25) * 100))


def _get_tier(score: int) -> str:
    if score >= 70:
        return "Core"
    if score >= 55:
        return "Strong"
    if score >= 40:
        return "Slow"
    return "Exit"


def score_products(products: list[dict], orders: list[dict], inventory: list[dict]) -> list[ScoredProduct]:
    """Score all products by revenue, velocity, stock health, and trend."""
    now = datetime.now(timezone.utc)
    seven_days_ago = now - timedelta(days=7)
    fourteen_days_ago = now - timedelta(days=14)

    # Aggregate sales per product
    product_sales: dict[str, dict] = {}
    for order in orders:
        if order.get("financial_status") == "refunded":
            continue
        order_time_str = order.get("processed_at") or order.get("created_at", "")
        try:
            order_time = datetime.fromisoformat(order_time_str.replace("Z", "+00:00"))
        except (ValueError, AttributeError):
            continue

        for item in (order.get("line_items") or []):
            # Match by title
            matched_product = None
            for p in products:
                if p.get("title") == item.get("title"):
                    matched_product = p
                    break
                variants = p.get("variants") or []
                if isinstance(variants, list):
                    for v in variants:
                        if isinstance(v, dict) and v.get("title") == item.get("variant_title"):
                            matched_product = p
                            break
                if matched_product:
                    break
            if not matched_product:
                continue

            pid = matched_product["id"]
            qty = item.get("quantity", 0)
            price = item.get("price", 0)
            if isinstance(price, str):
                price = float(price)

            sales = product_sales.setdefault(pid, {"units": 0, "revenue": 0.0, "recent": 0, "prior": 0})
            sales["units"] += qty
            sales["revenue"] += price * qty
            if order_time >= seven_days_ago:
                sales["recent"] += qty
            elif order_time >= fourteen_days_ago:
                sales["prior"] += qty

    # Inventory lookup
    inv_map: dict[str, int] = {}
    for level in inventory:
        pid = level.get("product_id", "")
        inv_map[pid] = inv_map.get(pid, 0) + level.get("quantity", 0)
    for p in products:
        if p["id"] not in inv_map:
            inv_map[p["id"]] = p.get("inventory_total", 0)

    # Max values for normalization
    max_revenue = max((s["revenue"] for s in product_sales.values()), default=1) or 1
    max_velocity = max((s["recent"] / 7 for s in product_sales.values()), default=1) or 1

    scored: list[ScoredProduct] = []
    for p in products:
        pid = p["id"]
        sales = product_sales.get(pid, {"units": 0, "revenue": 0.0, "recent": 0, "prior": 0})
        stock = inv_map.get(pid, 0)

        velocity = round(sales["recent"] / 7, 2)
        days_left = round(stock / velocity) if velocity > 0 else (999 if stock > 0 else 0)

        # Trend
        trend = "stable"
        trend_ratio = 1.0
        if sales["prior"] > 0:
            trend_ratio = round(sales["recent"] / sales["prior"], 2)
            if trend_ratio > 1.15:
                trend = "growing"
            elif trend_ratio < 0.85:
                trend = "declining"
        elif sales["recent"] > 0:
            trend = "growing"
            trend_ratio = 2.0

        # Score components
        rev_score = _power_scale(sales["revenue"], max_revenue)
        vel_score = _power_scale(velocity, max_velocity)
        stock_score = 0 if days_left <= 0 else 20 if days_left <= 3 else 50 if days_left <= 7 else 70 if days_left <= 14 else 90
        trend_score = min(100, max(0, round(50 + (trend_ratio - 1) * 25)))

        composite = round(rev_score * 0.3 + vel_score * 0.3 + stock_score * 0.2 + trend_score * 0.2)
        tier = _get_tier(composite)

        scored.append(ScoredProduct(
            id=pid, title=p.get("title", ""), handle=p.get("handle", ""),
            status=p.get("status", "active"), price=p.get("price_min", 0),
            inventory=stock, score=composite, tier=tier, velocity=velocity,
            days_left=days_left, trend=trend, trend_ratio=trend_ratio,
            revenue_total=round(sales["revenue"], 2),
            image=p.get("featured_image_url"),
        ))

    scored.sort(key=lambda s: (-1 if s.days_left <= 7 else 0, -s.score))
    return scored


def detect_stockout_risk(scored: list[ScoredProduct]) -> list[ScoredProduct]:
    return [p for p in scored if 0 < p.days_left <= 3 and p.velocity > 0]


def detect_slow_movers(scored: list[ScoredProduct]) -> list[ScoredProduct]:
    return [p for p in scored if p.trend == "declining" and p.inventory > 0 and p.days_left > 14 and p.tier != "Core"]


def check_product_health(products: list[dict], inventory: list[dict]) -> list[HealthIssue]:
    issues: list[HealthIssue] = []
    for p in products:
        if not p.get("featured_image_url"):
            issues.append(HealthIssue(p["id"], p.get("title", ""), "Missing product image", "warning"))
        if p.get("status") == "active" and p.get("inventory_total", 0) <= 0:
            issues.append(HealthIssue(p["id"], p.get("title", ""), "Active product with zero stock", "critical"))
        if p.get("status") == "draft" and p.get("inventory_total", 0) > 0:
            issues.append(HealthIssue(p["id"], p.get("title", ""), "Draft product has stock — consider publishing", "info"))
        if p.get("price_min", 0) <= 0 and p.get("status") == "active":
            issues.append(HealthIssue(p["id"], p.get("title", ""), "Active product with $0 price", "critical"))
    return issues


def suggest_discounts(slow_movers: list[ScoredProduct]) -> list[DiscountSuggestion]:
    suggestions: list[DiscountSuggestion] = []
    for product in slow_movers:
        pct = 10
        if product.trend_ratio < 0.5:
            pct = 30
        elif product.trend_ratio < 0.7:
            pct = 20
        elif product.trend_ratio < 0.85:
            pct = 15
        if product.tier == "Exit":
            pct = min(40, pct + 10)
        suggestions.append(DiscountSuggestion(product=product, discount_pct=pct))
    return suggestions

File: frontend/lib/intelligence.ts
/**
 * Intelligence Layer — Product scoring, velocity, stockout prediction, discount suggestions.
 *
 * Ported from Plus2's production scoring system (src/lib/products/scoring.ts).
 * Pure functions — no DB, no API calls, just math on arrays.
 */

import type { Product, Order, InventoryLevel } from './types'
import type { ScoredProduct, Tier, SegmentedCustomer, SegmentName } from './agents/types'
import type { Customer } from './types'

// ── Power Scale ─────────────────────────────────────────────────────────────
// Fourth-root scaling so $1K revenue scores ~50 while the top product scores 100.
// Sub-$100 products score low but the curve is gentler than square root.

function powerScale(value: number, maxValue: number): number {
  if (value <= 0 || maxValue <= 0) return 0
  return Math.min(100, Math.round(Math.pow(value / maxValue, 0.25) * 100))
}

function getTier(score: number): Tier {
  if (score >= 70) return 'Core'
  if (score >= 55) return 'Strong'
  if (score >= 40) return 'Slow'
  return 'Exit'
}

// ── Scoring ─────────────────────────────────────────────────────────────────

export function scoreProducts(
  products: Product[],
  orders: Order[],
  inventory: InventoryLevel[]
): ScoredProduct[] {
  // Build order line-item aggregates per product
  const productSales = new Map<string, { units: number; revenue: number; recentUnits: number; priorUnits: number }>()

  const now = Date.now()
  const sevenDaysAgo = now - 7 * 24 * 60 * 60 * 1000
  const fourteenDaysAgo = now - 14 * 24 * 60 * 60 * 1000

  for (const order of orders) {
    if (order.financial_status === 'refunded') continue
    const orderTime = new Date(order.processed_at || order.created_at).getTime()

    for (const item of order.line_items) {
      // Match line item to product by title (Shopify line items don't carry product_id)
      const product = products.find(
        (p) => p.title === item.title || p.variants.some((v) => v.title === item.variant_title)
      )
      if (!product) continue

      const existing = productSales.get(product.id) || { units: 0, revenue: 0, recentUnits: 0, priorUnits: 0 }
      existing.units += item.quantity
      existing.revenue += item.price * item.quantity

      if (orderTime >= sevenDaysAgo) {
        existing.recentUnits += item.quantity
      } else if (orderTime >= fourteenDaysAgo) {
        existing.priorUnits += item.quantity
      }

      productSales.set(product.id, existing)
    }
  }

  // Build inventory lookup (sum across variants)
  const inventoryMap = new Map<string, number>()
  for (const level of inventory) {
    const current = inventoryMap.get(level.product_id) || 0
    inventoryMap.set(level.product_id, current + level.quantity)
  }

  // Also sum from product.variants as fallback
  for (const product of products) {
    if (!inventoryMap.has(product.id)) {
      inventoryMap.set(product.id, product.inventory_total)
    }
  }

  // Pre-compute maximums for power-scale normalization
  const maxRevenue = Math.max(1, ...Array.from(productSales.values()).map((s) => s.revenue))
  const maxVelocity = Math.max(1, ...Array.from(productSales.values()).map((s) => s.recentUnits / 7))

  const scored: ScoredProduct[] = products.map((product) => {
    const sales = productSales.get(product.id) || { units: 0, revenue: 0, recentUnits: 0, priorUnits: 0 }
    const stock = inventoryMap.get(product.id) ?? product.inventory_total

    // Velocity: units sold per day over last 7 days
    const velocity = Math.round((sales.recentUnits / 7) * 100) / 100

    // Days until stockout
    const daysLeft = velocity > 0 ? Math.round(stock / velocity) : stock > 0 ? 999 : 0

    // Trend: compare last 7d vs prior 7d
    let trend: 'growing' | 'stable' | 'declining' = 'stable'
    let trendRatio = 1.0
    if (sales.priorUnits > 0) {
      trendRatio = Math.round((sales.recentUnits / sales.priorUnits) * 100) / 100
      if (trendRatio > 1.15) trend = 'growing'
      else if (trendRatio < 0.85) trend = 'declining'
    } else if (sales.recentUnits > 0) {
      trend = 'growing'
      trendRatio = 2.0
    }

    // Score components (each 0-100)
    const revenueScore = powerScale(sales.revenue, maxRevenue)       // 30%
    const velocityScore = powerScale(velocity, maxVelocity > 0 ? maxVelocity : 1)  // 30%
    const stockHealthScore = daysLeft <= 0 ? 0 : daysLeft <= 3 ? 20 : daysLeft <= 7 ? 50 : daysLeft <= 14 ? 70 : 90  // 20%
    const trendScore = Math.min(100, Math.max(0, Math.round(50 + (trendRatio - 1) * 25)))  // 20%

    const compositeScore = Math.round(
      revenueScore * 0.3 +
      velocityScore * 0.3 +
      stockHealthScore * 0.2 +
      trendScore * 0.2
    )

    const tier = getTier(compositeScore)

    return {
      id: product.id,
      title: product.title,
      handle: product.handle,
      status: product.status,
      price: product.price_min,
      inventory: stock,
      score: compositeScore,
      tier,
      velocity,
      daysLeft,
      trend,
      trendRatio,
      revenueTotal: Math.round(sales.revenue * 100) / 100,
      image: product.featured_image_url,
      variants: product.variants.map((v) => ({
        id: v.id,
        title: v.title,
        sku: v.sku,
        inventory_quantity: v.inventory_quantity,
        price: v.price,
      })),
    }
  })

  // Sort by days-until-stockout ascending (most urgent first), then by score descending
  scored.sort((a, b) => {
    if (a.daysLeft <= 7 && b.daysLeft > 7) return -1
    if (b.daysLeft <= 7 && a.daysLeft > 7) return 1
    return b.score - a.score
  })

  return scored
}

// ── Slow Mover Detection ────────────────────────────────────────────────────

export function detectSlowMovers(scored: ScoredProduct[]): ScoredProduct[] {
  return scored.filter(
    (p) => p.trend === 'declining' && p.inventory > 0 && p.daysLeft > 14 && p.tier !== 'Core'
  )
}

// ── Stockout Risk Detection ─────────────────────────────────────────────────

export function detectStockoutRisk(scored: ScoredProduct[]): ScoredProduct[] {
  return scored.filter((p) => p.daysLeft <= 3 && p.daysLeft > 0 && p.velocity > 0)
}

// ── Discount Suggestions ────────────────────────────────────────────────────

export function suggestDiscounts(slowMovers: ScoredProduct[]): { product: ScoredProduct; discountPct: number }[] {
  return slowMovers.map((product) => {
    // Deeper discount for worse trend + lower tier
    let pct = 10
    if (product.trendRatio < 0.5) pct = 30
    else if (product.trendRatio < 0.7) pct = 20
    else if (product.trendRatio < 0.85) pct = 15

    // Bump up for Exit tier
    if (product.tier === 'Exit') pct = Math.min(40, pct + 10)

    return { product, discountPct: pct }
  })
}

// ── Health Checks (from Plus2's shop-health.ts) ─────────────────────────────

export interface HealthIssue {
  productId: string
  productTitle: string
  issue: string
  severity: 'critical' | 'warning' | 'info'
}

export function checkProductHealth(products: Product[], inventory: InventoryLevel[]): HealthIssue[] {
  const issues: HealthIssue[] = []

  for (const product of products) {
    // Missing image
    if (!product.featured_image_url) {
      issues.push({
        productId: product.id,
        productTitle: product.title,
        issue: 'Missing product image',
        severity: 'warning',
      })
    }

    // Active with zero stock
    if (product.status === 'active' && product.inventory_total <= 0) {
      issues.push({
        productId: product.id,
        productTitle: product.title,
        issue: 'Active product with zero stock',
        severity: 'critical',
      })
    }

    // Draft with stock (should be published)
    if (product.status === 'draft' && product.inventory_total > 0) {
      issues.push({
        productId: product.id,
        productTitle: product.title,
        issue: 'Draft product has stock — consider publishing',
        severity: 'info',
      })
    }

    // Price at $0
    if (product.price_min <= 0 && product.status === 'active') {
      issues.push({
        productId: product.id,
        productTitle: product.title,
        issue: 'Active product with $0 price',
        severity: 'critical',
      })
    }
  }

  return issues
}

// ── RFM Customer Segmentation ───────────────────────────────────────────────

export function segmentCustomers(customers: Customer[], orders: Order[]): SegmentedCustomer[] {
  const now = Date.now()

  // Build per-customer aggregates from orders
  const customerOrders = new Map<string, { lastOrderAt: number; orderCount: number; totalSpent: number }>()

  for (const order of orders) {
    if (!order.customer_id || order.financial_status === 'refunded') continue
    const existing = customerOrders.get(order.customer_id) || { lastOrderAt: 0, orderCount: 0, totalSpent: 0 }
    const orderTime = new Date(order.processed_at || order.created_at).getTime()
    existing.lastOrderAt = Math.max(existing.lastOrderAt, orderTime)
    existing.orderCount += 1
    existing.totalSpent += order.total_price
    customerOrders.set(order.customer_id, existing)
  }

  // Compute RFM values
  const rfmData = customers.map((c) => {
    const agg = customerOrders.get(c.id) || {
      lastOrderAt: c.last_order_at ? new Date(c.last_order_at).getTime() : 0,
      orderCount: c.orders_count,
      totalSpent: c.total_spent,
    }
    const daysSinceLastOrder = agg.lastOrderAt > 0 ? Math.floor((now - agg.lastOrderAt) / (1000 * 60 * 60 * 24)) : 999
    return { customer: c, daysSinceLastOrder, orderCount: agg.orderCount, totalSpent: agg.totalSpent }
  })

  // Quintile scoring
  const recencyValues = rfmData.map((d) => d.daysSinceLastOrder).sort((a, b) => a - b)
  const frequencyValues = rfmData.map((d) => d.orderCount).sort((a, b) => a - b)
  const monetaryValues = rfmData.map((d) => d.totalSpent).sort((a, b) => a - b)

  function quintile(sortedVals: number[], value: number, invert = false): number {
    if (sortedVals.length === 0) return 3
    const idx = sortedVals.findIndex((v) => v >= value)
    const rank = idx === -1 ? sortedVals.length : idx
    const pct = rank / sortedVals.length
    const score = Math.min(5, Math.max(1, Math.ceil(pct * 5)))
    return invert ? 6 - score : score
  }

  return rfmData.map(({ customer, daysSinceLastOrder, orderCount, totalSpent }) => {
    // Recency: lower days = higher score (invert)
    const recencyScore = quintile(recencyValues, daysSinceLastOrder, true)
    const frequencyScore = quintile(frequencyValues, orderCount)
    const monetaryScore = quintile(monetaryValues, totalSpent)
    const rfmScore = Math.round((recencyScore + frequencyScore + monetaryScore) / 3 * 10) / 10

    let segment: SegmentName
    if (recencyScore >= 4 && frequencyScore >= 4 && monetaryScore >= 4) {
      segment = 'Champions'
    } else if (frequencyScore >= 3 && monetaryScore >= 3 && recencyScore >= 3) {
      segment = 'Loyal'
    } else if (recencyScore <= 2 && frequencyScore >= 3) {
      segment = 'At Risk'
    } else if (recencyScore >= 4 && frequencyScore <= 2) {
      segment = 'New'
    } else {
      segment = 'Lost'
    }

    return {
      id: customer.id,
      email: customer.email,
      name: `${customer.first_name} ${customer.last_name}`.trim(),
      segment,
      recencyScore,
      frequencyScore,
      monetaryScore,
      rfmScore,
      orderCount,
      totalSpent,
      daysSinceLastOrder,
    }
  })
}
